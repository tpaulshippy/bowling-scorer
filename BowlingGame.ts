import Frame from "./Frame";
import IBowler from './IBowler';
export default class BowlingGame implements IBowler {
    Frames: Frame[] = [];
    CurrentFrame: Frame = null;

    Hit(pins: number): void {
        if (!this.CurrentFrame || (this.CurrentFrame && this.CurrentFrame.IsClosed())) {
            if (this.Frames.length === 10 && this.CurrentFrame.IsClosed()) {
                throw new Error("game over");
            }

            this.CurrentFrame = new Frame(this.Frames.length);
            // Push current frame to array of frames
            this.Frames.push(this.CurrentFrame);
        }
        this.CurrentFrame.AddToFrame(pins);
    }

    Score(): number {
        let currentScore: number = 0;
        let framePlusOne: Frame = null;
        let framePlusTwo: Frame = null;

        for (let index = 0; index < this.Frames.length; index++) {
            const frame = this.Frames[index];

            if (frame.Index < 9) {
                framePlusOne = this.Frames[index + 1];
            }
            else {
                framePlusOne = null;
            }
            if (frame.Index < 8) {
                framePlusTwo = this.Frames[index + 2];
            }
            else {
                framePlusTwo = null;
            }

            if (frame.Hits && frame.Hits.length > 1 && frame.TotalPins() === 10) {
                // Spare
                if (framePlusOne) {
                    frame.ScoreBonus.push(framePlusOne.Hits[0]);
                }
            }

            if (frame.Hits[0] === 10) {
                // Strike
                if (framePlusOne) {
                    frame.ScoreBonus.push(framePlusOne.Hits[0]);
                    if (framePlusOne.Hits.length > 1) {
                        frame.ScoreBonus.push(framePlusOne.Hits[1]);
                    }
                    else if (framePlusTwo) {
                        frame.ScoreBonus.push(framePlusTwo.Hits[0]);
                    }
                }
            }
            currentScore += frame.Score();
        }


        return currentScore;
    }


}

