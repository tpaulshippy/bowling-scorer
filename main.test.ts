import BowlingGame from './BowlingGame';
it("scores perfect game", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl twelve strikes
    for (let i = 0; i < 12; i++) {
        game.Hit(10);
    }

    // THEN
    // Expect score of 300
    const score = game.Score();
    expect(score).toEqual(300);
    checkBonuses(game);

    expect(game).toMatchSnapshot();

});
it("scores all gutter balls", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl twenty gutter balls
    for (let i = 0; i < 20; i++) {
        game.Hit(0);
    }

    // THEN
    // Expect score of 0
    const score = game.Score();
    expect(score).toEqual(0);
    checkBonuses(game);

});
it("scores all nines and spares", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl ten spares
    for (let i = 0; i < 10; i++) {
        game.Hit(9);
        game.Hit(1);
    }
    // Bowl extra shot from 10th frame
    game.Hit(9);

    // THEN
    // Expect score of 19 * 10 = 190
    const score = game.Score();
    expect(score).toEqual(190);
    checkBonuses(game);

});

it("scores strike and then all nines and spares", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl strike
    game.Hit(10);

    // Bowl nine spares
    for (let i = 0; i < 9; i++) {
        game.Hit(9);
        game.Hit(1);
    }
    // Bowl extra shot from 10th frame
    game.Hit(9);

    // THEN
    // Expect score of 20 + (19 * 9) = 191
    const score = game.Score();
    expect(score).toEqual(191);
    checkBonuses(game);

});
it("scores two strikes and then all nines and spares", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl two strikes
    game.Hit(10);
    game.Hit(10);

    // Bowl eight spares
    for (let i = 0; i < 8; i++) {
        game.Hit(9);
        game.Hit(1);
    }
    // Bowl extra shot from 10th frame
    game.Hit(9);

    // THEN
    // Expect score of 29 + 20 + (19 * 8) = 201
    const score = game.Score();
    expect(score).toEqual(201);
    checkBonuses(game);
    expect(game).toMatchSnapshot();
});

it("scores all nines and spares except three strikes in tenth", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl ten spares
    for (let i = 0; i < 9; i++) {
        game.Hit(9);
        game.Hit(1);
    }
    // Bowl turkey
    game.Hit(10);
    game.Hit(10);
    game.Hit(10);

    // THEN
    // Expect score of 19 * 8 = 152 + 20 + 30 = 202
    const score = game.Score();
    expect(score).toEqual(202);
    checkBonuses(game);

});

it("scores all nines", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl ten nines
    for (let i = 0; i < 10; i++) {
        game.Hit(9);
        game.Hit(0); // second ball is gutter ball
    }

    // THEN
    // Expect score of 9 * 10 = 90
    const score = game.Score();
    expect(score).toEqual(90);
    checkBonuses(game);

});


it("scores 0 - 10", () => {
    // GIVEN
    // Initialize game
    const game = new BowlingGame();

    // WHEN
    // Bowl ten times starting with a gutter ball and then each time hitting one more ball
    for (let i = 0; i < 10; i++) {
        game.Hit(i);
        game.Hit(0); // second ball is gutter ball
    }

    // THEN
    // Expect score of 0 + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9
    const score = game.Score();
    expect(score).toEqual(0 + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9);
    checkBonuses(game);
});

function checkBonuses(game: BowlingGame) {
    // Ensure that spares have one bonus and strikes have two
    for (const frame of game.Frames) {
        if (frame.Hits[0] === 10 && frame.Index < 9) {
            expect(frame.ScoreBonus.length).toEqual(2);
        }
        else if (frame.TotalPins() === 10 && frame.Index < 9) {
            expect(frame.ScoreBonus.length).toEqual(1);
        }
    }
}