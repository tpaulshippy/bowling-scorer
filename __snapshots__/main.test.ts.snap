// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scores perfect game 1`] = `
BowlingGame {
  "CurrentFrame": Frame {
    "Hits": Array [
      10,
      10,
      10,
    ],
    "Index": 9,
    "ScoreBonus": Array [],
  },
  "Frames": Array [
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 0,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 1,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 2,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 3,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 4,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 5,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 6,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 7,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 8,
      "ScoreBonus": Array [
        10,
        10,
      ],
    },
    Frame {
      "Hits": Array [
        10,
        10,
        10,
      ],
      "Index": 9,
      "ScoreBonus": Array [],
    },
  ],
}
`;

exports[`scores two strikes and then all nines and spares 1`] = `
BowlingGame {
  "CurrentFrame": Frame {
    "Hits": Array [
      9,
      1,
      9,
    ],
    "Index": 9,
    "ScoreBonus": Array [],
  },
  "Frames": Array [
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 0,
      "ScoreBonus": Array [
        10,
        9,
      ],
    },
    Frame {
      "Hits": Array [
        10,
      ],
      "Index": 1,
      "ScoreBonus": Array [
        9,
        1,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 2,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 3,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 4,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 5,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 6,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 7,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
      ],
      "Index": 8,
      "ScoreBonus": Array [
        9,
      ],
    },
    Frame {
      "Hits": Array [
        9,
        1,
        9,
      ],
      "Index": 9,
      "ScoreBonus": Array [],
    },
  ],
}
`;
